### PR TITLE
[prelude] make Stream lazier

### PR DESCRIPTION
**Pull request by hearnadam**
- stop using class member, enforcing accidental eagerness
- add new method: Stream.range to simplify creation of streams
- init/range: accept `chunkSize` to limit size of Chunks- Chunk.from (Array): remove AnyRef type bound

The first PR to address #679. This change is important to ensure the laziness of streams initialized from non-effectful sources.

### CODE REVIEW

File: kyo-prelude/shared/src/main/scala/kyo/Stream.scala [hunk 1]
```diff
@@ -328,25 +329,52 @@ final case class Stream[V, -S](v: Ack < (Emit[Chunk[V]] & S)):
 end Stream
 
 object Stream:
+    @nowarn("msg=anonymous")
+    inline def apply[V, S](v: => Ack < (Emit[Chunk[V]] & S)): Stream[V, S] =
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/Stream.scala:333]: we can avoid the function allocation for `v`:

```suggestion
    inline def apply[V, S](inline v: => Ack < (Emit[Chunk[V]] & S)): Stream[V, S] =
```

File: kyo-prelude/shared/src/main/scala/kyo/Stream.scala [hunk 2]
```diff
@@ -328,25 +329,52 @@ final case class Stream[V, -S](v: Ack < (Emit[Chunk[V]] & S)):
...
+                        if c.isEmpty then Loop.done(Ack.Continue())
+                        else
+                            val i = n min chunkSize
+                            Emit.andMap(c.take(i))(ack => Loop.continue(c.dropLeft(i), ack))
+            }
+        }
+
+    private def fromIterable[V, S](iterable: Iterable[V], chunkSize: Int)(using tag: Tag[Emit[Chunk[V]]], frame: Frame) =
+        Emit.andMap(Chunk.empty[V]) { ack =>
+            val iterator = iterable.iterator
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/Stream.scala:364]: I'm not sure about this since using the iterator is a side effect

Comment (hearnadam) [reply kyo-prelude/shared/src/main/scala/kyo/Stream.scala]: I removed it for now. Let's see if it's necessary later.

File: kyo-prelude/shared/src/main/scala/kyo/Stream.scala [hunk 3]
```diff
@@ -328,25 +329,90 @@ final case class Stream[V, -S](v: Ack < (Emit[Chunk[V]] & S)):
...
+    @nowarn("msg=anonymous")
+    inline def apply[V, S](inline v: => Ack < (Emit[Chunk[V]] & S)): Stream[V, S] =
+        new Stream[V, S]:
+            def emit: Ack < (Emit[Chunk[V]] & S) = v
 
     private val _empty           = Stream(Stop)
     def empty[V]: Stream[V, Any] = _empty.asInstanceOf[Stream[V, Any]]
 
-    def init[V, S](seq: Seq[V] < S)(using tag: Tag[Emit[Chunk[V]]], frame: Frame): Stream[V, S] =
+    private inline def DefaultChunkSize = 4096
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/Stream.scala:340]: What do you think about making this public? It'd be nice to document it and allow users to inspect it.

File: kyo-prelude/shared/src/main/scala/kyo/Stream.scala [hunk 4]
```diff
@@ -328,25 +329,90 @@ final case class Stream[V, -S](v: Ack < (Emit[Chunk[V]] & S)):
...
+
+                                if !continue then Loop.done(Stop)
+                                else
+                                    val remaining =
+                                        if step > 0 then
+                                            ((end - current - 1) / step).abs + 1
+                                        else
+                                            ((current - end - 1) / step.abs).abs + 1
+                                    val size  = (n min _chunkSize) min remaining
+                                    val chunk = Chunk.from(Array.range(current, current + size * step, step))
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/Stream.scala:411]: Could we use `collection.immutable.Range` here? It represents a range without allocating the individual items.

**Review by fwbrasil — APPROVED** [review]

**Pull request merged**